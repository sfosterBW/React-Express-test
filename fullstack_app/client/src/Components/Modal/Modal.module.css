.wrapper {
  position: absolute;
  bottom: 0;
  right: 0;
  left: 0;
  top: 0;
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal {
  position: relative;
  border-radius: 5px;
  background: lightgray;
}

.modal form {
  background: lightgray;
}

.title {
  font-size: 3vmin;
  text-align: center;
}

.close {
  position: absolute;
  top: 1.5vmin;
  right: 0px;
  transform: translateY(-50%);
  cursor: pointer;
}
