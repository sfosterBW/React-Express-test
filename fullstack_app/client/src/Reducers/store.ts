import { applyMiddleware, combineReducers, createStore } from "redux"
import thunkMiddleware from "redux-thunk"
import { composeWithDevTools } from 'redux-devtools-extension'
import modalReducer from './modalReducer'
import alertReducer from './alertReducer'
import fruitReducer from './fruitReducer'

export const rootReducer = combineReducers({
  alert: alertReducer,
  fruit: fruitReducer,
  modal: modalReducer
})

export type RootState = ReturnType<typeof rootReducer>

export const configureStore = (preloadedState: undefined) => {
  const middlewares = [thunkMiddleware]
  const middlewareEnhancer = applyMiddleware(...middlewares)
  const enhancers = [middlewareEnhancer]
  const composedEnhancers = composeWithDevTools(...enhancers)
  const store = createStore(rootReducer, preloadedState, composedEnhancers)
  return store
}

export const store = configureStore(undefined)

export default configureStore
