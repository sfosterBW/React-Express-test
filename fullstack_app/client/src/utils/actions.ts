import { IFruit } from './interfaces'
 
//Action const
export const TOGGLE_ALERT = "TOGGLE_ALERT"

//Action shape
interface ToggleAlert {
  type: typeof TOGGLE_ALERT,
  payload: boolean
}

//Action type
export type AlertActionTypes = ToggleAlert

//Action
export function toggleAlert(alertValue: boolean): AlertActionTypes {
  return {
    type: TOGGLE_ALERT,
    payload: alertValue
  }
}

//Action const
export const TOGGLE_MODAL = "TOGGLE_MODAL"
export const OPEN_MODAL = "OPEN_MODAL"

//Action shape
interface ToggleModal {
  type: typeof TOGGLE_MODAL,
  payload: boolean
}

interface OpenModal {
  type: typeof OPEN_MODAL,
  payload: IFruit
}

//Action type
export type ModalActionTypes = ToggleModal | OpenModal

//Action
export function toggleModal(modalValue: boolean): ModalActionTypes {
  return {
    type: TOGGLE_MODAL,
    payload: modalValue,
  }
}

export function openModal(fruit: IFruit): ModalActionTypes {
  return {
    type: OPEN_MODAL,
    payload: fruit,
  }
}

//Action const
export const GET_FRUITS = 'GET_FRUITS'
export const CREATE_FRUIT = 'CREATE_FRUIT'
export const REMOVE_FRUIT = 'REMOVE_FRUIT'
export const UPDATE_FRUIT = 'UPDATE_FRUIT'

//Action shape
interface GetFruits {
  type: typeof GET_FRUITS,
  payload: IFruit[]
}

interface CreateFruit {
  type: typeof CREATE_FRUIT,
  payload: IFruit
}

interface RemoveFruit {
  type: typeof REMOVE_FRUIT,
  payload: number
}

interface UpdateFruit {
  type: typeof UPDATE_FRUIT,
  payload: IFruit
}

//Action types
export type FruitActionTypes = GetFruits | CreateFruit | RemoveFruit | UpdateFruit

//Actions
export const getFruits = (fruits: IFruit[]): FruitActionTypes => {
  return {
    type: GET_FRUITS,
    payload: fruits
  }
}

export const createFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: CREATE_FRUIT,
    payload: fruit,
  }
}

export const removeFruit = (id: number): FruitActionTypes => {
  return {
    type: REMOVE_FRUIT,
    payload: id,
  }
}

export const updateFruit = (fruit: IFruit): FruitActionTypes => {
  return {
    type: UPDATE_FRUIT,
    payload: fruit,
  }
}
