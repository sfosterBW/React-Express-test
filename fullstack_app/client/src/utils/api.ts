import axios from 'axios'
import { IFruit } from './interfaces'
 
axios.defaults.baseURL = '/fruit-api/'

const fetchFruit = async () => {
  const response = await axios.get("list")
  return response.data
}

const createFruit = async (fruit: { name: string, best: boolean }) => {
    const response = await axios.post('new', { new: fruit })
    return response.data
}

const updateFruit = async (fruit: IFruit) => {
    const response = await axios.put('update', fruit)
    return response.data
}

const deleteFruit = async (id: number) => {
  const deleteData = { params: { id: id } }
  const response = await axios.delete('delete', deleteData)
  return response.data
}

export default { fetchFruit, createFruit, updateFruit, deleteFruit }
