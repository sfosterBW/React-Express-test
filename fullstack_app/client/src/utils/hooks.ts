import { useState } from 'react'

export const useField = (
  initValue: string,
  label: string,
  name: string,
  type: string) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any) => {
    setValue(event.target.value)
  }

  const reset = () => {
    setValue("")
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
