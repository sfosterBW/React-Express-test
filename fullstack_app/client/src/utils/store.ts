import { applyMiddleware, combineReducers, createStore } from "redux"
import thunkMiddleware from "redux-thunk"
import { composeWithDevTools } from 'redux-devtools-extension'
import {
  alertReducer,
  fruitReducer,
  modalReducer } from './reducers'

export const rootReducer = combineReducers({
  alert: alertReducer,
  fruit: fruitReducer,
  modal: modalReducer
})

export type RootState = ReturnType<typeof rootReducer>

export default function configureStore(preloadedState: undefined) {
  const middlewares = [thunkMiddleware]
  const middlewareEnhancer = applyMiddleware(...middlewares)
  const enhancers = [middlewareEnhancer]
  const composedEnhancers = composeWithDevTools(...enhancers)
  const store = createStore(rootReducer, preloadedState, composedEnhancers)
  return store
}

export const store = configureStore(undefined)
