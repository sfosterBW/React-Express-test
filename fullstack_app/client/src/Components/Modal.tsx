import React, { FC } from 'react'
import { RootState } from '../utils/store'
import { useSelector, useDispatch } from 'react-redux'
import { toggleModal } from '../utils/actions'
import Form from './Form'
import { IFruit } from '../utils/interfaces'

interface Props {
  fruit: IFruit | undefined
  title: string
}

const Modal: FC<Props> = ({ fruit, title }) => {

  const selectToggle = (state: RootState) => state.modal.toggle
  const toggle = useSelector(selectToggle)
  const dispatch = useDispatch()

  const displayModal = () => {
    if (toggle && fruit) {
      return (
        <div className="modal-wrapper">
          <div className="modal">
            <h2 className="modal-title">{title}</h2>
            <Form fruit={fruit} />
            <button
              className="closeButton"
              name="close"
              onClick={() => dispatch(toggleModal(!toggle))}>
              &times;
              </button>
          </div>
        </div>
      )
    } else {
      console.log("Modal fruit undefined")
      return null
    }
  }

  return (
    <div>{displayModal()}</div>
  )
}

export default Modal
