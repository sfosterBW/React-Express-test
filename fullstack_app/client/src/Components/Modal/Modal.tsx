import React, { FC } from 'react'
import { RootState } from '../../utils/store'
import { useSelector, useDispatch } from 'react-redux'
import { toggleModal } from '../../utils/actions'
import { IFruit } from '../../utils/interfaces'
import Form from '../Form/Form'
import styles from './Modal.module.css'

interface Props {
  fruit: IFruit | undefined
  title: string
}

const Modal: FC<Props> = ({ fruit, title }) => {

  const selectToggle = (state: RootState) => state.modal.toggle
  const toggle = useSelector(selectToggle)
  const dispatch = useDispatch()

  const displayModal = () => {
    if (toggle && fruit) {
      return (
        <section className={styles.wrapper}>
          <div className={styles.modal}>
            <h2 className={styles.title}>{title}</h2>
            <Form fruit={fruit} />
            <button
              className={styles.close}
              name="close"
              onClick={() => dispatch(toggleModal(!toggle))}>
              &times;
              </button>
          </div>
        </section>
      )
    }
  }

  return (
    <div>{displayModal()}</div>
  )
}

export default Modal
