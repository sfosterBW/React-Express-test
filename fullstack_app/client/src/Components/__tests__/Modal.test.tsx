
import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { createStore } from 'redux'

import Modal from '../Modal'
import { rootReducer } from '../../utils/store'

describe('the Modal component', () => {

  const modalValue = true
  const mockStore = createStore(rootReducer, { modal: { toggle: modalValue } })
  const mockFunction = jest.fn()
  const mockFruit = {
    _id: 1,
    name: "False Case",
    best: false
  }
  const title = "This is a title"
  const modalComponent =
    <Provider store={mockStore}>
      <Modal fruit={mockFruit} handleSubmit={mockFunction} title={title} />
    </Provider>

  const modal = mount(modalComponent)

  it('renders with the correct structure', () => {
    expect(modal).toBeDefined()
    expect(modal.find('div')).toHaveLength(5)
    expect(modal.find('div.modal-wrapper')).toHaveLength(1)
    expect(modal.find('div.modal')).toHaveLength(1)
    expect(modal.find('h2.modal-title')).toHaveLength(1)
    expect(modal.find('h2.modal-title').text()).toEqual(title)
    expect(modal.find('form')).toHaveLength(1)
    expect(modal.find('button.closeButton')).toHaveLength(1)
  })

  it('functions as expected', () => {
    modal.find('button.closeButton').simulate('click')
    expect(modal.find('div')).toHaveLength(1)
  })

  it('renders the same as last time', () => {
    const tree = renderer
      .create(modalComponent)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
