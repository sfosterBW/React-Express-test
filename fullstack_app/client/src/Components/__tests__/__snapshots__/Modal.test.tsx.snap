// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Modal component renders the same as last time 1`] = `
<div>
  <div
    className="modal-wrapper"
  >
    <div
      className="modal"
    >
      <h2
        className="modal-title"
      >
        This is a title
      </h2>
      <form
        className="fruit-form"
        onSubmit={[Function]}
      >
        <h2>
          List out your favourite fruit
        </h2>
        <div
          className="input-wrapper"
        >
          <label
            htmlFor="name"
            id="name"
          >
            Add a fruit:
          </label>
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value="False Case"
          />
        </div>
        <div
          className="input-wrapper"
        >
          <label
            htmlFor="best"
            id="best"
          >
            Is it best?:
          </label>
          <input
            checked={false}
            name="best"
            onChange={[Function]}
            type="checkbox"
          />
        </div>
        <button>
          Add new fruit
        </button>
      </form>
      <button
        className="closeButton"
        name="close"
        onClick={[Function]}
      >
        Ã—
      </button>
    </div>
  </div>
</div>
`;
